python/pyspark/sql/types.py

```diff
@@ -599,6 +599,8 @@ def toInternal(self, obj):
             if isinstance(obj, dict):
                 return tuple(f.toInternal(obj.get(n)) if c else obj.get(n)
                              for n, f, c in zip(self.names, self.fields, self._needConversion))
+            elif isinstance(obj, Row):
+                return self.toInternal(obj.asDict())
```

Comment:
Hm, `Row` inherits `tuple` so this is supposed to be handled below.

Reply:
yes, but because `Row` also behaves like a `dict` (as in `row[colName]`) the tuple handling can be off. e.g. if you have `row = Row(name='Alice', date=date(2019,10,1))` the underlying tuple reverses the order of fields (see examples above).

How about

`elif isinstance(obj, Row) and getattr(obj, "__from_dict__", False):`

as in https://github.com/apache/spark/blob/master/python/pyspark/sql/types.py#L614

?

Reply:
I added this change. At least in `types.py` there were two places where the same distinction was necessary.

I also added a test to check that RDD-style `Row`s are still processed correctly.

Reply:
I added the `and getattr(obj, "__from_dict__", False)`.

In `types.py` there were two places where the same distinction was necessary.

I also added a test to check that RDD-style `Row`s are still processed correctly.

Reply:
@HyukjinKwon Anything else I should address?